src/Lumper/BSP/JsonContractResolver.cs

```diff
@@ -0,0 +1,46 @@
+﻿using System;
+using System.Linq;
+using System.Collections.Generic;
+using Lumper.Lib.BSP.Lumps;
+using Lumper.Lib.BSP.IO;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Serialization;
+
+namespace Lumper.Lib.BSP
+{
+    public class JsonContractResolver : DefaultContractResolver
+    {
+        public bool SortProperties { get; }
+        public bool IgnoreOffset { get; }
+
+        public JsonContractResolver(bool sortProperties, bool ignoreOffset)
+        {
+            SortProperties = sortProperties;
+            IgnoreOffset = ignoreOffset;
+        }
+
+        protected override IList<JsonProperty> CreateProperties(Type type,
+            MemberSerialization memberSerialization)
+        {
+            var @base = base.CreateProperties(type, memberSerialization);
+            IEnumerable<JsonProperty> ordered = @base;
+            if (SortProperties)
+                ordered = ordered.OrderBy(p => p.Order ?? int.MaxValue)
+                                 .ThenBy(p => p.PropertyName);
+            if (IgnoreOffset)
+            {
+                ordered = ordered.Where(x =>
+                    !x.DeclaringType.IsAssignableTo(typeof(IUnmanagedLump))
+                    || x.PropertyName != nameof(IUnmanagedLump.DataStreamOffset));
+                ordered = ordered.Where(x =>
+                    (
+                       !x.DeclaringType.IsAssignableFrom(typeof(LumpHeader))
+                    && !x.DeclaringType.IsAssignableFrom(typeof(BspLumpHeader))
+                    && !x.DeclaringType.IsAssignableFrom(typeof(GameLumpHeader))
+                    )
+                    || x.PropertyName != nameof(LumpHeader.Offset));
+            }
+            return ordered.ToList();
```

Comment:
Really don't like this nesting, early returning here is better even if it means writing `ToList()` twice.

Reply:
this BIRD likes nesting .. and I don't like multiple returns .. and I don't like redundant code. I usually just change this stuff because I don't like arguing about it, but I really don't see the benefit in changing this.

Reply:
Okay fair enough, looking back at this you're totally right, it'd make the logic a lot less clear.

Though, I do find this block genuinely really hard to read. The indentation is pretty idiosyncratic but whatever, eventually we'll add some formatting rules that will keep everything consistent. I think the real problem is that the logic is just totally over-complicated. I had to copy this over to a text editor to pick it apart, and can't it just simplify down to this?

```c#
if (IgnoreOffset)
{
  ordered = ordered.Where(x =>
    !(x.DeclaringType.IsAssignableTo(typeof(IUnmanagedLump))
      || x.DeclaringType.IsAssignableFrom(typeof(LumpHeader))
      || x.DeclaringType.IsAssignableFrom(typeof(BspLumpHeader))
      || x.DeclaringType.IsAssignableFrom(typeof(GameLumpHeader))
      || x.PropertyName == nameof(IUnmanagedLump.DataStreamOffset)
      || x.PropertyName == nameof(LumpHeader.Offset)
    )
}
return ordered.ToList();
```

Reply:
the weird formatting is supposed to make it easier to read .. you code doesn't do the same thing .. how about I invert the entire thing instead?
```C#
                ordered = ordered.Where(x =>
                    !(x.DeclaringType.IsAssignableTo(typeof(IUnmanagedLump))
                    && x.PropertyName == nameof(IUnmanagedLump.DataStreamOffset)));
                ordered = ordered.Where(x =>
                    !((
                       x.DeclaringType.IsAssignableFrom(typeof(LumpHeader))
                    || x.DeclaringType.IsAssignableFrom(typeof(BspLumpHeader))
                    || x.DeclaringType.IsAssignableFrom(typeof(GameLumpHeader))
                    )
                    && x.PropertyName == nameof(LumpHeader.Offset)));
```

Reply:
Ohhhhh I'm sorry, I flipped the top one around wrong. But definitely using the DeMorgan rule or whatever like that makes it easy to follow. In terms of formatting, I just don't find the specific indentation you've done helpful, since you have to first deduce the reasoning behind the formatting method before you can tell what it's denoting. Again, with consistent formatting that'd be fine, but reviewing this I keep having to figure stuff out case-by-case.

Reply:
Sorry, to clarify, let's use the change you suggested